src/Compilers/Core/Portable/Collections/CachingDictionary.cs

```diff
@@ -237,7 +237,7 @@ private static bool IsNotFullyPopulatedMap([NotNullWhen(returnValue: false)] IDi
 
                     if (!existingMap.TryGetValue(key, out elements))
                     {
-                        elements = _getElementsOfKey(key);
+                        elements = existingMap.GetOrAdd(key, _getElementsOfKey(key));
```

Comment:
It looks like GetOrAdd has an overload which could take `_getElementsOfKey` and `key` and only call the delegate if it is necessary